audit policy: support subresources and resource names

ericchiang opened the pull request:
Updates #48561

	policy:
	- level: Metadata
	  resources:
	  - group: ""
	    resources ["pods/logs"]
	- level: None
	  resources:
	  - group: ""
	    resources: ["configmaps"]
	    resourceNames: ["controller-leader"]

The top level resource no longer matches the subresource. For example "pods"
no longer matches requests to the logs subresource on pods.

```release-note
Audit policy supports matching subresources and resource names, but the top level resource no longer matches the subresouce. For example "pods" no longer matches requests to the logs subresource of pods. Use "pods/logs" to match subresources.
```

/cc @sttts @soltysh @crassirostris @ihmccreery @timstclair

tallclair commented:
> Subresources uses a "(resource)/(subresource)" format for similarity
with RBAC. This is a breaking change since specifying the resource
alone (e.g. "pods") would previously match all subresources on the
resource (e.g. "pods/logs").

please make this a release note

tallclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,10 +200,17 @@ type GroupResources struct {
+	// Resources is a list of resources within the API group. Specific subresources
+	// must be matched using a "/" to indicate the subresource. For example "pods" does
+	// not match "pods/logs".
+	// An empty list implies every resource kind in the API group.
```
Do we want to add a wildcard for subresource matching? E.g. `pods/*` matches all pod subresources? I'm not sure whether this would be useful.

Alternatively, is matching all subresources the common case? In that case, we might want to make `pods` match `pods/logs`, and require another rule to match `pods/logs` separately if desired.

WDYT?

ericchiang replied:
> Alternatively, is matching all subresources the common case? In that case, we might want to make pods match pods/logs, and require another rule to match pods/logs separately if desired.

I like this better.

deads2k replied:
> Do we want to add a wildcard for subresource matching? E.g. pods/* matches all pod subresources? I'm not sure whether this would be useful.

We have talked about adding this sort of `*` capability as an additive change to RBAC in support of polymorphic endpoints like `*/scale`.  Normally, `resource` *must not* allow `resource/subresource` since the subresources provide more power.  `pods/bind`, `pods/status`, `namespaces/finalize`, or any other status subresource.

@kubernetes/sig-auth-misc @liggitt

tallclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,10 +200,17 @@ type GroupResources struct {
+	// Using this field requires Resources to be specified.
+	// An empty list implies every resource instead of the resource.
+	// +optional
+	ResourceNames []string
```
Would it be better to specify resource names per-resource? It's easier to write out the profile yaml this way, but more precise (and probably more accurately reflects the desired state) to attach the names to the resource types.

tallclair replied:
Then again, if I need to specify `pods`, `pods/exec`, and `pods/logs`, etc. separately, it would be nice to just list the pod name once.

ericchiang replied:
I'd prefer to keep this API closer to RBAC, which specifies ResourceNames as unique values.

tallclair commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go:
```diff
@@ -159,4 +187,7 @@ func TestChecker(t *testing.T) {
 	test("nonResource", audit.LevelRequestResponse, "getLogs", "getClusterRoles", "default")
+	test("nonResource", audit.LevelNone, "getPodLogs", "getPods")
+
+	test("subresource", audit.LevelRequest, "getPodLogs", "getPods")
```
Test that `subresource` doesn't match `getPods`.

tallclair reviewed (commented):
Thanks for tackling this! A couple open questions though.

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -143,14 +143,29 @@ func ruleMatchesResource(r *audit.PolicyRule, attrs authorizer.Attributes) bool
+				// 		"pods" matches "pods" and "pods/logs"
+				//		"pods/logs" only matches "pods/logs"
+				//
+				if r == resource || (fullResource != "" && r == fullResource) {
```
so we cannot match the non-subresource alone? We could use "pods/" for that.

ericchiang replied:
I figured that rules aren't purely additive would be enough. In contrast to RBAC, you can do something like this:

```yaml
policy:
- level: None
  resources: ["pods/level"]
- level: Metadata
  resources: ["pods"]
```

I think having `pods/` only match the non-subresourced component would be fine too though.

sttts replied:
Have no strong opinion. Your example is also fine. Just make sure the comments are correct.

tallclair replied:
Let's consider the user perspective. A new user is trying to set up audit logging. The want to log pod requests, so they do:
```yaml
policy:
- level: Request
  resources: ["pods"]
```
Do you think they would expect a request to `pods/status` to also be logged? I think the answer is probably yes...

tallclair replied:
Either way, we should make sure that it's documented.

ericchiang replied:
Added a note to the top level comment. Will update kubernetes.io docs when this gets in.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -201,10 +201,17 @@ type GroupResources struct {
 	Resources []string `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// ResourceNames is a list of resource instance names that the policy matches.
+	// Using this field requires Resources to be specified.
+	// An empty list implies every resource instead of the resource.
```
"every resource instead of the resource" is a little difficult to understand.

ericchiang replied:
Maybe

```
An empty list implies that ever instance of the resource is matched.
```

ericchiang commented:
/test pull-kubernetes-federation-e2e-gce

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go:
```diff
@@ -88,16 +88,18 @@ func validateResources(groupResources []audit.GroupResources, fldPath *field.Pat
-		if msgs := validation.NameIsDNSSubdomain(groupResource.Group, false); len(msgs) != 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), groupResource.Group, strings.Join(msgs, ",")))
+		if len(groupResource.ResourceNames) > 0 && len(groupResource.Resources) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), groupResource.ResourceNames, "using ResourceNames requires at least one Resource"))
```
nit: fldPath.Child("group") --> fldPath.Child("resourceNames")

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,10 +200,27 @@ type GroupResources struct {
+	//		- level: None
+	// 		  resources: ["pods/logs"] # Ignore pods/logs.
+	// 		- level: Metadata
+	//		  resources: ["pods"] # Catch all for "pods" and all other pods subresources.
```
Nit
the policy example is invalid, It should be:
```
- level: None
  resources: ["pods/log"] # Ignore pods/log.
- level: Metadata
   resources: ["pods"] # Catch all for "pods" and all other pods subresources.
```

1. a middle level "resources" is missed
2. subresource is "log" rather than "logs"

We  also need to update the example in github page, I think.
https://github.com/kubernetes/kubernetes/pull/48836#issue-242523068

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -201,10 +201,27 @@ type GroupResources struct {
+	//		- level: None
+	// 		  resources: ["pods/logs"] # Ignore pods/logs.
+	// 		- level: Metadata
+	//		  resources: ["pods"] # Catch all for "pods" and all other pods subresources.
```
ditto

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -143,14 +143,29 @@ func ruleMatchesResource(r *audit.PolicyRule, attrs authorizer.Attributes) bool
+				// Resource either matches the resource directly or matches the
+				// resource + the subresource.
+				//
+				// 		"pods" matches "pods" and "pods/logs"
```
s/log/logs/

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go:
```diff
@@ -61,6 +61,18 @@ func TestChecker(t *testing.T) {
+			APIGroup:        "", // Core
+			APIVersion:      "v1",
+			Resource:        "pods",
+			Subresource:     "logs",
```
s/logs/log/

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.proto:
```diff
@@ -106,10 +106,28 @@ message GroupResources {
+  // 		- level: None
+  // 		  resources: ["pods/logs"] # Ignore pods/logs.
+  // 		- level: Metadata
+  // 		  resources: ["pods"] # Catch all for "pods" and all other pods subresources.
```
ditto.

CaoShuFeng commented:
This patch test OK in my envrionment.
Thanks.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,10 +200,27 @@ type GroupResources struct {
+	//
+	//		policy:
+	//		- level: None
+	// 		  resources: ["pods/log"] # Ignore pods/log.
```
Policy format in this example is illegal.
The allowed format which is acceptable to --audit-policy-file option is:
```
  - level: None
    resources:
      - resources: ["pods/log"] # Ignore pods/log.
  - level: Metadata
    resources:
      - resources: ["pods"] # Catch all for "pods" and all other pods subresources.
```
Please take look at the 2nd and 5th line.

CaoShuFeng commented:
/test pull-kubernetes-verify
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-node-e2e
/test pull-kubernetes-unit

ericchiang commented:
(having some issues with the protobuf generation scripts, give me a sec to try to rectify that)

sttts commented:
@ericchiang is this blocked on the protobuf topic or anything else?

ericchiang commented:
@sttts I actually don't think anythings blocking this. Any outstanding comments? I'm not seeing any from the thread.

/test pull-kubernetes-e2e-gce-etcd3

crassirostris commented:
I have one comment @sttts mentioned to me recently. @CaoShuFeng in https://github.com/kubernetes/kubernetes/pull/49115 moved the API to v1beta1

Maybe we should merge it first and then your changes will be applied to the v1beta1 API?

ericchiang commented:
@crassirostris yeah makes sense. I'll re-apply the changes once #49115 is merged.

CaoShuFeng commented:
#49115 is merged。
Sorry that my pr makes you redo some work. :(

ericchiang committed:
Audit policy v1beta1 now supports matching subresources and resource names.

	policy:
	- level: Metadata
	  resources:
	  - group: ""
	    resources ["pods/logs"]
	- level: None
	  resources:
	  - group: ""
	    resources: ["configmaps"]
	    resourceNames: ["controller-leader"]

The top level resource no longer matches the subresource. For example "pods"
no longer matches requests to the logs subresource on pods.

```release-note
Audit policy supports matching subresources and resource names, but the top level resource no longer matches the subresouce. For example "pods" no longer matches requests to the logs subresource of pods. Use "pods/logs" to match subresources.
```

ericchiang commented:
PR updated to target the v1beta1 API group. No other changes.

ericchiang commented:
Hitting a generation issue. It looks like `./hack/update-codegen.sh` is creating an invalid generated go file.

```
$ ./hack/update-codegen.sh 
make: Entering directory '/home/eric/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/eric/src/k8s.io/kubernetes'
# ...
$ git status
On branch audit-policy-subresource-resource-name
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go
	modified:   staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/zz_generated.conversion.go
	modified:   staging/src/k8s.io/apiserver/pkg/apis/audit/zz_generated.deepcopy.go

no changes added to commit (use "git add" and/or "git commit -a")
$ go install -v ./vendor/k8s.io/apiserver/pkg/apis/audit/...
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/fuzzer
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1beta1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/validation
# k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1
vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go:46:3: undefined: Convert_audit_GroupResources_To_v1alpha1_GroupResources
vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go:305:14: undefined: Convert_audit_GroupResources_To_v1alpha1_GroupResources
```

For some reason the script is removing one of the generated methods in the conversion file

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go b/staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go
index 972dca99f0..8f2a74eecf 100644
--- a/staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go
+++ b/staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/zz_generated.conversion.go
@@ -147,14 +147,10 @@ func Convert_v1alpha1_GroupResources_To_audit_GroupResources(in *GroupResources,
 func autoConvert_audit_GroupResources_To_v1alpha1_GroupResources(in *audit.GroupResources, out *GroupResources, s conversion.Scope) error {
        out.Group = in.Group
        out.Resources = *(*[]string)(unsafe.Pointer(&in.Resources))
+       // WARNING: in.ResourceNames requires manual conversion: does not exist in peer-type
        return nil
 }
 
-// Convert_audit_GroupResources_To_v1alpha1_GroupResources is an autogenerated conversion function.
-func Convert_audit_GroupResources_To_v1alpha1_GroupResources(in *audit.GroupResources, out *GroupResources, s conversion.Scope) error {
-       return autoConvert_audit_GroupResources_To_v1alpha1_GroupResources(in, out, s)
-}
-
 func autoConvert_v1alpha1_ObjectReference_To_audit_ObjectReference(in *ObjectReference, out *audit.ObjectReference, s conversion.Scope) error {
        out.Resource = in.Resource
        out.Namespace = in.Namespace
```

Will keep looking into this, but anyone have any ideas?

ericchiang committed:
generated: update API resources

	./hack/update-codegen.sh
	./hack/update-codecgen.sh
	./hack/update-generated-protobuf.sh

ericchiang commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,30 @@
+	"k8s.io/apiserver/pkg/apis/audit"
+)
+
+// Convert_audit_GroupResources_To_v1alpha1_GroupResources is a manually created conversion
```
Manual conversion added. Tried to copy the RBAC conversion https://github.com/kubernetes/kubernetes/blob/v1.7.4/pkg/apis/rbac/v1alpha1/conversion.go

ericchiang replied:
hmmm, this now fails fuzzing. @sttts @CaoShuFeng any advice here?

CaoShuFeng replied:
I thinks this fails because ResourceName is lost when we decode Policy back.
Can we always set ResourceName to empty string in fuzzer.go?  @sttts

liggitt replied:
RBAC was able to derive or default the fields that were different between v1alpha1 and v1beta1... it didn't just drop a field. I'd actually have expected the new field to have been added to v1alpha1 as well for simplicity.

ericchiang replied:
> I'd actually have expected the new field to have been added to v1alpha1 as well for simplicity.

Just going to do this.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go:
```diff
@@ -88,6 +100,11 @@ func TestChecker(t *testing.T) {
+		"getPodLogs": {
+			Level:     audit.LevelRequest,
+			Verbs:     []string{"get"},
+			Resources: []audit.GroupResources{{Resources: []string{"pods/logs"}}},
```
This break one unit test because "pods/logs" doesn't not consistent with line 71.

ericchiang replied:
Yeah I've got the change locally. Will push once I figure out the fuzzing

crassirostris commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -215,10 +215,31 @@ type GroupResources struct {
 	Resources []string
+	// ResourceNames is a list of resource instance names that the policy matches.
+	// Using this field requires Resources to be specified.
+	// An empty list implies that ever instance of the resource is matched.
```
everY

ericchiang replied:
thanks :+1:  updated.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -143,14 +143,29 @@ func ruleMatchesResource(r *audit.PolicyRule, attrs authorizer.Attributes) bool
 
 	apiGroup := attrs.GetAPIGroup()
 	resource := attrs.GetResource()
+	fullResource := ""
```
It's not `fullResource`, it's `subresource`, since it's empty when there's no subresource

crassirostris commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -143,14 +143,29 @@ func ruleMatchesResource(r *audit.PolicyRule, attrs authorizer.Attributes) bool
+				// 		"pods" matches "pods" and "pods/log"
+				//		"pods/log" only matches "pods/log"
+				//
+				if r == resource || (fullResource != "" && r == fullResource) {
```
- I think it would benefit to the readability to replace nested ifs with consequent
- Can resource name be empty?

```
if r != resource && r != subresource {
  continue
}
if len(gr.ResourceNames) == 0 || hasString(gr.ResourceNames, name) {
  return true
}
```

ericchiang replied:
Switched this so `fullResource == resource` when there's no subresource. It adds a redundant check, but the code's a little easier to follow. ptal

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -220,6 +220,11 @@ type GroupResources struct {
 	// Any empty list implies every resource kind in the API group.
 	// +optional
 	Resources []string `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
```
Does the v1alpha1 Resources support the "resource/subresource" logic?
Do we need to update docstring for Resources?

ericchiang replied:
It does for 1.8, but won't on a 1.7 cluster. I'll expand this comment to mention that uses who want that behavior should just switch to the beta group.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -143,14 +143,29 @@ func ruleMatchesResource(r *audit.PolicyRule, attrs authorizer.Attributes) bool
-				if res == resource {
-					return true
+			for _, r := range gr.Resources {
+				// Resource either matches the resource directly or matches the
```
I think it's worth to specify which resource is which, e.g. `Resource from the policy rule either matches the target resource directly or target resource + subresource`. WDYT?

crassirostris reviewed (commented):
LGTM, one nit

ericchiang commented:
/retest

ericchiang commented:
/retest

I think all outstanding comments are addressed btw.

CaoShuFeng commented:
LGTM

ericchiang commented:
/assign @sttts

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -215,10 +215,31 @@ type GroupResources struct {
-	// Any empty list implies every resource kind in the API group.
+	// Resources is a list of resources within the API group. Specific subresources
+	// can be matched using a "/" to indicate the subresource. For example "pods" matches
+	// requests to "pods" and "pods/log" but "pods/log" only matches "pods/log".
```
And pods/ matches only the main resource?

ericchiang replied:
pods/ doesn't match anything :)

sttts replied:
Maybe it should? I only want to audit spec updates, not millions of status updates. How?

crassirostris replied:
I see a valid use case for each of the following

- auditing all subresources
- auditing only the main resource
- auditing only the subset of subresources

Maybe we should create a separate API object to fully express all cases? E.g.

```go
type SubresourcePolicyRuleType int

const (
  Whitelist SubresourcePolicyRuleType = iota
  Blacklist
)

type SubresourcePolicyRule struct {
  Type SubresourcePolicyRuleType
  Subresources []string
}
```

or the same fields inlined in the `PolicyRule`. Checking this will be easier also

crassirostris replied:
...or just match `pods/` with the main resource, that's also reasonable

sttts replied:
That was my expectation: `pods` for all, `pods/` for the main resource, `pods/log` for log. I don't think we need big subresource resource for that.

liggitt replied:
that seems odd to me... I'd expect `pods` to just be the main resource, and something involving a `/` (`pods/*`?) to be all subresources

ericchiang replied:
I think switching back to the RBAC implementation where "pods" only matches the top level resource is reasonable. In the discussion above we made "pods" match all subresources because that seemed like a reasonable default, but it actually doesn't seem like it is.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -216,10 +216,23 @@ type GroupResources struct {
 	// Any empty list implies every resource kind in the API group.
 	// +optional
 	Resources []string `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// The ResourceNames field was added for parody with the v1beta1 API group and is unsupported
```
This is odd. Why do we have this field? In an alpha API group we can of course easily add anything. But what does "unsupported" mean?

The official way to add a field in a beta is to either
a) add it to the alpha too (what we do here)
b) store it in annotations for alpha during conversion.

With a) we completely support the new field also in the alpha.

ericchiang replied:
"unsupported" tries to convey that some API servers that support the v1alpha1 API group don't actually support this field (e.g. if you use a 1.7 server).

Might have overthought this one. Let me just add it to the v1alpha1 API group normally.

sttts replied:
I would still use b) in general. Here for alpha I don't care. But adding everything new to old apigroups feels wrong.

crassirostris replied:
+1 to b)

liggitt replied:
doing a field -> annotation conversion means custom conversions... what is the benefit over adding the field? let's not spend more effort on v1alpha1 than required

sttts replied:
@liggitt that's what I meant with "for alpha I do not care". we will drop that anyway soon. Let's leave this additional field, but remove the strange comment about "not supported".

ericchiang replied:
Will remove the comment

ericchiang commented:
Based on discussion going to change this so the top level resource doesn't match subresources. If you specify "pods" you won't match "pods/logs", you have to explicitly provide "pods/logs" in your policy.

I think this is reasonable behavior given how different subresources behave. e.g. "/status" subresources get updated consistently even though they don't modify the declared state of the cluster. We don't want to match these by default.

In the future we can add conveniences like "pods/*", but we need to make the breaking change now to stop matching subresources before we go to beta.

cc @sttts @tallclair @liggitt @crassirostris @CaoShuFeng

crassirostris commented:
SGTM

ericchiang commented:
/retest

sttts commented:
/retest

sttts commented:
/lgtm
/approve no-issue

sttts commented:
/retest

soltysh commented:
/retest

ericchiang commented:
/retest
